sql: implement generic indexed variables for expressions

RaduBerinde opened the pull request:
Implementing generic support for expression variables that are addressed using
an index. This replaces the `scanQValues` and will also be used by the new
TableReader.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6316)

<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[sql/parser/indexed_vars.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6316#-KGJBzxuXKA_9NhJdKsy:-KGJBzxuXKA_9NhJdKsz:-1594310130) ([raw file](https://github.com/cockroachdb/cockroach/blob/c66372114612838968ab403bc967a533ab702568/sql/parser/indexed_vars.go#L28)):*
There is a bit of stuttering that is un-go-like with `IndexedVar` as the prefix for all of these methods, but not having that prefix would be confusing. I don't have a suggestion, though.

---

*[sql/parser/indexed_vars.go, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6316#-KGJBu1rHOaaRHaibqDf:-KGJBu1s2qzHVxLuOHes:1482874601) ([raw file](https://github.com/cockroachdb/cockroach/blob/c66372114612838968ab403bc967a533ab702568/sql/parser/indexed_vars.go#L74)):*
These could be methods on `IndexedVarHelper` instead of `*IndexedVarHelper`, though I don't think that matters here.

---

*[sql/parser/indexed_vars.go, line 94 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6316#-KGJ5ikr8RBCKaXjal4a:-KGJ5ikr8RBCKaXjal4b:2056791480) ([raw file](https://github.com/cockroachdb/cockroach/blob/c66372114612838968ab403bc967a533ab702568/sql/parser/indexed_vars.go#L94)):*
I don't think this needs to be exported.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6316#-:-KGJCRXUZlf8hjY3ucXi:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

---

Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[sql/parser/indexed_vars.go, line 94 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6316#-KGJ5ikr8RBCKaXjal4a:-KGJE44xPv9O4PIkzAZe:959408502) ([raw file](https://github.com/cockroachdb/cockroach/blob/c66372114612838968ab403bc967a533ab702568/sql/parser/indexed_vars.go#L94)):*
I will need it for the table reader.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6316#-:-KGJDvJ2VIRuKqT_bVAS:-485616692)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql: implement generic indexed variables for expressions

Implementing generic support for expression variables that are addressed using
an index. This replaces the `scanQValues` and will also be used by the new
TableReader.
